lib/IDE/Refactoring.cpp

```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
                                           EditConsumer).performChange();
 }
 
+static std::unique_ptr<llvm::SetVector<Expr*>> findConcatenatedExpressions(ResolvedRangeInfo Info) {
+  if (Info.Kind != RangeKind::SingleExpression && Info.Kind != RangeKind::PartOfExpression)
+    return nullptr;
+  Expr *E = Info.ContainedNodes[0].get<Expr*>();
+
+  struct StringInterpolationExprFinder: public SourceEntityWalker {
+    std::unique_ptr<llvm::SetVector<Expr*>> Bucket = llvm::make_unique<llvm::SetVector<Expr*>>();
+    bool IsValidInterpolation = true;
+    StringInterpolationExprFinder() {}
+
+    bool isConcatenationExpr(DeclRefExpr* Expr) {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD->getName().str() != "+") {
```

Comment:
We should compare decl pointer instead of its name for "+" here. Could you try to add a function like "getStringDecl()" in ASTContext and use it here?

Reply:
Hmm... Could you help me a bit on this? I've changed the other string comparisons but this one is a bit more tricky. What I've done is: I've found `KnownDecls.def` which seems to define functions and is a potential place to dig in. I could add a new entry and as I see there is a macro that generated concrete implementation of the getter. But how should I name it(string literal form)? How does binding look like between this declaration and the object in memory? I haven't found any examples for using `FUNCDECL` macro with operators like `+`. Thanks in advance!

Reply:
I think we can use the function I added in #11964. Please let me know if that works.

Reply:
Thanks for this! Unfortunately `FD` and `ASTContext getPlusFunctionOnRangeReplaceableCollection()` are not the same. That's weird since String from Swift 4 adopts this protocol. I'm using `FD` like in the example above. Should I maybe refer to it in some other way?

Reply:
Hm, it's defined directly in `String.swift` here: https://github.com/apple/swift/blob/master/stdlib/public/core/String.swift#L873 It's probably why it doesn't work.

Reply:
I've ended up adding new function on `ASTContext`